scripts/soft-reboot

```diff
@@ -0,0 +1,217 @@
+    elif grep -q onie_platform= /host/machine.conf; then
+        KERNEL_OPTIONS=$(cat /host/grub/grub.cfg | sed "/$NEXT_SONIC_IMAGE'/,/}/"'!'"g" | grep linux)
+        KERNEL_IMAGE="/host$(echo $KERNEL_OPTIONS | cut -d ' ' -f 2)"
+        BOOT_OPTIONS="$(echo $KERNEL_OPTIONS | sed -e 's/\s*linux\s*/BOOT_IMAGE=/') SONIC_BOOT_TYPE=${BOOT_TYPE_ARG}"
+    else
+        error "Unknown bootloader. ${REBOOT_TYPE} is not supported."
+        exit "${EXIT_NOT_SUPPORTED}"
+    fi
+    INITRD=$(echo $KERNEL_IMAGE | sed 's/vmlinuz/initrd.img/g')
+}
+
+function load_kernel() {
+    # Load kernel into the memory
+    /sbin/kexec -l "$KERNEL_IMAGE" --initrd="$INITRD" --append="$BOOT_OPTIONS"
+}
+
+function reboot_pre_check()
+{
+    # Make sure that the file system is normal: read-write able
+    filename="/host/test-`date +%Y%m%d-%H%M%S`"
+    ERR=0
+    touch ${filename} || ERR=$?
+    if [[ ${ERR} -ne 0 ]]; then
+        # Continue rebooting in this case, but log the error
+        VERBOSE=yes debug "Filesystem might be read-only or full ..."
+    fi
+    rm ${filename}
+
+    # Verify the next image by sonic-installer
+    local message=$(sonic-installer verify-next-image 2>&1)
+    if [ $? -ne 0 ]; then
+        VERBOSE=yes debug "Failed to verify next image: ${message}"
+        exit ${EXIT_SONIC_INSTALLER_VERIFY_REBOOT}
+    fi
+}
+
+function parse_options()
+{
+    while getopts "h?v" opt; do
+        case ${opt} in
+            h|\? )
+                show_help_and_exit
+                ;;
+            v )
+                VERBOSE=yes
+                ;;
+            t )
+                TAG_LATEST=no
+                ;;
+        esac
+    done
+}
+
+parse_options $@
+
+# Exit if not superuser
+if [[ "$EUID" -ne 0 ]]; then
+    echo "This command must be run as root" >&2
+    exit "${EXIT_FAILURE}"
+fi
+
+if [ -x ${LOG_SSD_HEALTH} ]; then
+    debug "Collecting logs to check ssd health before ${REBOOT_TYPE}..."
+    ${LOG_SSD_HEALTH}
+fi
+
+debug "User requested rebooting device ..."
+
+setup_reboot_variables
+reboot_pre_check
+
+# Tag remotely deployed images as local
+tag_images
+
+# Stop SONiC services gracefully.
+stop_sonic_services
+
+clear_warm_boot
+
+echo "$KERNEL_IMAGE : $INITRD : $BOOT_OPTIONS"
+
+load_kernel
+
+# Update the reboot cause file to reflect that user issued 'reboot' command
+# Upon next boot, the contents of this file will be used to determine the
+# cause of the previous reboot
+echo "User issued '${REBOOT_SCRIPT_NAME}' command [User: ${REBOOT_USER}, Time: ${REBOOT_TIME}]" > ${REBOOT_CAUSE_FILE}
+
+sync
+sleep 3
+sync
+
+# sync the current system time to CMOS
+if [ -x /sbin/hwclock ]; then
+    /sbin/hwclock -w || /bin/true
+fi
+
+# Run platform specific reboot plugin
+if [ -x ${DEVPATH}/${PLATFORM}/${PLATFORM_PLUGIN} ]; then
+    debug "Running ${PLATFORM} specific plugin..."
+    ${DEVPATH}/${PLATFORM}/${PLATFORM_PLUGIN}
+fi
+
```

Comment:
Do we want to also invoke watchdog here?

Reply:
This is a good question. But since this reboot is to prevent from performing the hardware reboot on those specific devices, we would be better to notify the reboot failure and exit this case. 
Or we can add some condition to check the s6100 specific ssd module gpio status before invoking the watchdog. 
What do you think?

Reply:
I discussed with Guohan about the watchdog arming before calling platform_plugin and decided to add it here.